src/mpid/ch4/shm/ipc/src/ipc_p2p.h

```diff
@@ -185,34 +185,32 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_IPCI_handle_lmt_recv(MPIDI_IPC_hdr * ipc_hdr,
         MPIR_ERR_CHECK(mpi_errno);
         /* skip unmap */
     } else if (ipc_hdr->ipc_type == MPIDI_IPCI_TYPE__GPU) {
-        /* try fast memcpy first */
-        int fast_copy = 0;
-        if (src_data_sz <= MPIR_CVAR_CH4_IPC_GPU_FAST_COPY_MAX_SIZE) {
-            MPIDI_Datatype_check_contig(MPIDIG_REQUEST(rreq, datatype), dt_contig);
-            if (ipc_hdr->is_contig && dt_contig) {
-                mpi_errno =
-                    MPIDI_GPU_ipc_fast_memcpy(ipc_hdr->ipc_handle, MPIDIG_REQUEST(rreq, buffer),
-                                              src_data_sz, MPIDIG_REQUEST(rreq, datatype));
-                if (mpi_errno == MPI_SUCCESS)
-                    fast_copy = 1;
-            }
-        }
-        if (!fast_copy) {
-            MPL_pointer_attr_t attr;
-            MPIR_GPU_query_pointer_attr(MPIDIG_REQUEST(rreq, buffer), &attr);
-            int dev_id = MPL_gpu_get_dev_id_from_attr(&attr);
-            int map_dev = MPIDI_GPU_ipc_get_map_dev(ipc_hdr->ipc_handle.gpu.global_dev_id, dev_id,
-                                                    MPIDIG_REQUEST(rreq, datatype));
-            mpi_errno = MPIDI_GPU_ipc_handle_map(ipc_hdr->ipc_handle.gpu, map_dev, &src_buf, false);
+        bool do_mmap = (src_data_sz <= MPIR_CVAR_CH4_IPC_GPU_FAST_COPY_MAX_SIZE);
+        MPL_pointer_attr_t attr;
+        MPIR_GPU_query_pointer_attr(MPIDIG_REQUEST(rreq, buffer), &attr);
+        MPIDI_Datatype_check_contig(MPIDIG_REQUEST(rreq, datatype), dt_contig);
+        int dev_id = MPL_gpu_get_dev_id_from_attr(&attr);
+        int map_dev = MPIDI_GPU_ipc_get_map_dev(ipc_hdr->ipc_handle.gpu.global_dev_id, dev_id,
+                                                MPIDIG_REQUEST(rreq, datatype));
+        mpi_errno = MPIDI_GPU_ipc_handle_map(ipc_hdr->ipc_handle.gpu, map_dev, &src_buf, do_mmap);
+        MPIR_ERR_CHECK(mpi_errno);
+        /* copy */
+        if (ipc_hdr->is_contig && dt_contig) {
+            mpi_errno = MPIR_Localcopy_gpu(src_buf, src_data_sz, MPI_BYTE, NULL,
```

Comment:
`MPIR_Localcopy_gpu` is failing the CUDA test. I guess that means it is buggy.

Reply:
I believe it is because of that lack of implementation for `MPL_gpu_imemcpy`. The macro guards should hopefully resolve these failures